test/Makefile

```diff
@@ -0,0 +1,95 @@
+TARGET_EXEC := canlib-tests
+
+BUILD_DIR := ./build
+SRCS_DIR := ../src/src tests/src
+
+SRCS := $(shell find $(SRC_DIRS) -name *.cpp -or -name *.c)
+
+COMP_SRC := ../src/src/computers/*.cpp
+
+# String substitution for every C/C++ file.
+# As an example, hello.cpp turns into ./build/hello.cpp.o
+
+# String substitution (suffix version without %).
+# As an example, ./build/hello.cpp.o turns into ./build/hello.cpp.d
+#DEPS := $(OBJS:.o=.d)
+
+# Every folder in ./src will need to be passed to GCC so that it can find header files
+INC_DIRS := $(shell find ../src/inc tests/inc -type d)
+# Add a prefix to INC_DIRS. So moduleA would become -ImoduleA. GCC understands this -I flag
+INC_FLAGS := -I../src/inc -Itests/inc
+COMPUTER_FLAGS := -I../src/inc/drivers -I../src/inc/computers
+
+# The -MMD and -MP flags together generate Makefiles for us!
+# These files will have .d instead of .o as the output.
+CPPFLAGS := $(INC_FLAGS) -std=c++17
+
+# The final build step.
+gen:
+	python3 ../generator/main.py can_spec.yml --testing
+	
+main0: gen
+	g++ $(CPPFLAGS) $(COMPUTER_FLAGS) $(CXXFLAGS) -DCANLIB_ARCH_TESTFAMILY -DARCH0 ../src/src/bus.cpp ../src/src/structs.cpp ../src/src/drivers/testfamily.cpp ../src/src/pack_unpack.cpp ../src/src/computers/canlib_testcomp0.cpp tests/src/identify.cpp tests/src/message.cpp tests/src/send.cpp tests/src/constants.cpp tests/src/arch0.cpp tests/main.cpp -o main0
+main1: gen
+	g++ $(CPPFLAGS) $(COMPUTER_FLAGS) $(CXXFLAGS) -DCANLIB_ARCH_TESTFAMILY -DARCH1 ../src/src/bus.cpp ../src/src/structs.cpp ../src/src/drivers/testfamily.cpp ../src/src/pack_unpack.cpp ../src/src/computers/canlib_testcomp1.cpp tests/src/identify.cpp tests/src/message.cpp tests/src/send.cpp tests/src/constants.cpp tests/src/arch1.cpp tests/main.cpp -o main1
+main2: gen
+	g++ $(CPPFLAGS) $(COMPUTER_FLAGS) $(CXXFLAGS) -DCANLIB_ARCH_TESTFAMILY -DARCH2 ../src/src/bus.cpp ../src/src/structs.cpp ../src/src/drivers/testfamily.cpp ../src/src/pack_unpack.cpp ../src/src/computers/canlib_testcomp2.cpp tests/src/identify.cpp tests/src/message.cpp tests/src/send.cpp tests/src/constants.cpp tests/src/arch2.cpp tests/main.cpp -o main2
+
+$(BUILD_DIR):
+	mkdir -p build/src/src
+	mkdir -p build/src/drivers
+	mkdir -p build/src/src/computers
+	mkdir -p build/tests/src
+
+# Build step for C++ source
+$(BUILD_DIR)/src/src/bus.cpp.o: $(BUILD_DIR)
+	g++ ../src/src/bus.cpp $(CPPFLAGS) $(CXXFLAGS) -c $< -o $(BUILD_DIR)/src/src/bus.cpp.o
+
+$(BUILD_DIR)/src/src/structs.cpp.o: $(BUILD_DIR)
+	g++ ../src/src/structs.cpp $(CPPFLAGS) $(CXXFLAGS) -c $< -o $(BUILD_DIR)/src/src/structs.cpp.o
+
+$(BUILD_DIR)/src/src/pack_unpack.cpp.o: $(BUILD_DIR)
+	g++ ../src/src/pack_unpack.cpp $(CPPFLAGS) $(CXXFLAGS) -c $< -o $(BUILD_DIR)/src/src/pack_unpack.cpp.o
```

Comment:
You could use a pattern matching rule for `.cpp` files instead of manually specifying each one.  This also seems to mean the object files are not dependent on the source files.  I would look at the MY20 Makefile rules.